File: js/arrivals.js
import { CONFIG } from "./config.js";
import { addLog } from "./state.js";
import { actIndex, ACT_ORDER, storyHook } from "./story.js";
import { countOccupants, getShelterStock } from "./shelter.js";
import {
  answerFor,
  availableQuestions,
  pickTiedStory,
} from "./media.js";

/**
 * The desk decision.
 *
 * Turning someone away only means something if the player has reasons, so every
 * arrival is weighed on three axes that all pull against each other:
 *
 *   capacity  — is there a bed at all
 *   sustain   — one more mouth costs measurable days of water and food
 *   threat    — evidence that this person is not what they claim
 *
 * And refusing is never free: sending away someone genuine costs humanity.
 * Evidence is deliberately probabilistic. Honest people have odd habits and
 * dangerous ones can look ordinary, so this stays a judgement call instead of a
 * lookup table.
 */

/** What the person actually is. The player never sees this directly. */
export const KIND = {
  traveler: "traveler",
  survivor: "survivor",
  wrong: "wrong",
};

const CLAIMS = {
  traveler: [
    "Driving through. Needs a bed and an early checkout.",
    "Says the next motel is another two hours and they are done driving.",
    "Here for work in the city. Paying by the night.",
    "Wants the quietest room you have. Does not say why.",
  ],
  survivor: [
    "Walked here. Asks how many people are already inside.",
    "Came from the east side. Will not talk about what they saw.",
    "Has a child with them. Asks only for somewhere with a door.",
    "Offers to work for a bunk. No money left.",
  ],
  wrong: [
    "Says they have a reservation. There is no reservation.",
    "Asks how many people are inside. Asks again, differently.",
    "Says they came from the city. The roads from the city are closed.",
    "Wants a room on the ground floor, near the back.",
  ],
};

/** Signs that genuinely indicate something is wrong. */
const DAMNING_SIGNS = [
  "Their reflection in the lobby glass lags a half-second behind them.",
  "They sign the register with the wrong year and do not correct it.",
  "They repeat your last words back to you, quietly, before answering.",
  "They do not blink while you explain the checkout time.",
  "Their bag is empty. You can tell by the way it hangs.",
  "They know Mary's name. Mary has not said it.",
  "They are not breathing when they think you are not looking.",
];

/** Signs that mean nothing at all. Honest people are strange too. */
const INNOCUOUS_SIGNS = [
  "They are dressed for winter. It is not winter.",
  "They ask for a room facing away from the road.",
  "They pay entirely in coins.",
  "They flinch at the ice machine.",
  "They will not put their bag down.",
  "They ask twice whether the doors lock from the inside.",
  "They keep checking the window behind you.",
  "They have not slept. It shows.",
];

const GENERIC_QUESTIONS = [
  {
    id: "where-from",
    prompt: "Where did you come from?",
    answers: {
      traveler: "Down the access road. I have been driving since morning.",
      survivor: "On foot. I do not want to say the last town out loud.",
      wrong: "The city. Everyone comes from the city.",
    },
  },
  {
    id: "how-long",
    prompt: "How long do you plan to stay?",
    answers: {
      traveler: "One night. Maybe two if the roads are still bad.",
      survivor: "Until it is safe. I do not know when that is.",
      wrong: "As long as you will let me. I can stay in any room.",
    },
  },
];

/** Hours burned by pressing an arrival for more information. */
const QUESTION_HOURS = 0.75;

function pick(list) {
  return list[Math.floor(Math.random() * list.length)];
}

function pickKind(state) {
  const act = actIndex(state);
  const roll = Math.random();

  // Act 1-2: paying travelers, with a rare wrong one to seed the doubt.
  if (act <= ACT_ORDER.indexOf("unease")) {
    if (roll < (act === 0 ? 0.05 : 0.14)) return KIND.wrong;
    return KIND.traveler;
  }
  // Act 3: money still works, but people are starting to just need shelter.
  if (act === ACT_ORDER.indexOf("disruption")) {
    if (roll < 0.24) return KIND.wrong;
    if (roll < 0.62) return KIND.survivor;
    return KIND.traveler;
  }
  // Act 4+: nobody is travelling for fun any more.
  if (roll < 0.3) return KIND.wrong;
  return KIND.survivor;
}

function buildSigns(kind) {
  const signs = [];
  const damningPool = [...DAMNING_SIGNS];
  const innocuousPool = [...INNOCUOUS_SIGNS];

  const take = (pool, damning) => {
    if (!pool.length) return;
    const index = Math.floor(Math.random() * pool.length);
    const [text] = pool.splice(index, 1);
    signs.push({ text, damning, revealed: false });
  };

  if (kind === KIND.wrong) {
    // Mostly real tells, plus noise so the read is never trivial.
    const damningCount = 1 + Math.floor(Math.random() * 3);
    for (let i = 0; i < damningCount; i++) take(damningPool, true);
    if (Math.random() < 0.5) take(innocuousPool, false);
  } else {
    const innocuousCount = Math.floor(Math.random() * 3);
    for (let i = 0; i < innocuousCount; i++) take(innocuousPool, false);
    // False positives: honest people occasionally look damning.
    if (Math.random() < 0.16) take(damningPool, true);
  }

  // Shuffle so damning signs are not always revealed first.
  signs.sort(() => Math.random() - 0.5);
  if (signs.length) signs[0].revealed = true;
  return signs;
}

export function createArrival(state, name) {
  const kind = pickKind(state);
  const story = pickTiedStory(state, kind);
  return {
    name,
    kind,
    storyId: story?.id ?? null,
    claim: pick(CLAIMS[kind]),
    signs: buildSigns(kind),
    questionsAsked: 0,
    maxQuestions: 2,
    askedQuestionIds: [],
    waitRemainingHours: CONFIG.waitPatienceHours,
    /** Kept for story hooks that only care that something was off. */
    marked: kind !== KIND.traveler,
  };
}

/** Questions the player can still put to this arrival. */
export function deskQuestions(state, guest) {
  const mediaQs = availableQuestions(state, guest);
  const asked = new Set(guest?.askedQuestionIds ?? []);
  const generic = GENERIC_QUESTIONS.filter((q) => !asked.has(q.id));
  const unusedMedia = mediaQs.filter((q) => !asked.has(`${q.storyId}:${q.id}`));
  return [...unusedMedia, ...generic];
}

export function revealedSigns(guest) {
  return (guest?.signs ?? []).filter((s) => s.revealed);
}

export function hiddenSignCount(guest) {
  return (guest?.signs ?? []).filter((s) => !s.revealed).length;
}

/** True if the player has been shown at least one genuine tell. */
export function hasVisibleTell(guest) {
  return revealedSigns(guest).some((s) => s.damning);
}

/**
 * The "why" panel: everything the player weighs before choosing.
 * Numbers only appear once the resource actually matters.
 */
export function assessArrival(state, guest) {
  const bunksFree = state.rooms.filter(
    (r) => r.unlocked && r.status === "clean"
  ).length;
  const bunksTotal = state.rooms.filter((r) => r.unlocked).length;

  const assessment = {
    bunksFree,
    bunksTotal,
    occupants: countOccupants(state),
    signsRevealed: revealedSigns(guest).length,
    questionsLeft: Math.max(0, guest.maxQuestions - guest.questionsAsked),
    shelter: null,
    humanity: state.story?.humanity ?? 100,
    paysRent: guest.kind === KIND.traveler,
  };

  if (state.shelter?.unlocked) {
    const use = CONFIG.shelterUse;
    const now = assessment.occupants;
    const after = now + 1;
    const days = (stock, perPerson) =>
      perPerson > 0 ? Math.floor(stock / (perPerson * Math.max(1, now))) : 0;
    const daysAfter = (stock, perPerson) =>
      perPerson > 0 ? Math.floor(stock / (perPerson * Math.max(1, after))) : 0;

    assessment.shelter = {
      waterDays: days(getShelterStock(state, "water"), use.waterPerPersonPerDay),
      waterDaysAfter: daysAfter(
        getShelterStock(state, "water"),
        use.waterPerPersonPerDay
      ),
      foodDays: days(getShelterStock(state, "food"), use.foodPerPersonPerDay),
      foodDaysAfter: daysAfter(
        getShelterStock(state, "food"),
        use.foodPerPersonPerDay
      ),
      integrity: Math.round(state.shelter.integrity),
      powered: state.shelter.powered,
    };
  }

  return assessment;
}

/**
 * Press for more information. Costs time.
 * If a media question is chosen, their answer is the tell — radio answers
 * are public knowledge, paper answers are not.
 */
export function askArrivalQuestion(state, guest, question = null) {
  if (!guest) return false;
  if (guest.questionsAsked >= guest.maxQuestions) {
    addLog(state, `${guest.name} stops answering questions.`);
    return false;
  }

  const options = deskQuestions(state, guest);
  const chosen = question ?? options[0] ?? GENERIC_QUESTIONS[0];
  guest.questionsAsked += 1;
  guest.waitRemainingHours -= QUESTION_HOURS;
  state.hour += QUESTION_HOURS;
  guest.askedQuestionIds = guest.askedQuestionIds || [];
  guest.askedQuestionIds.push(
    chosen.storyId ? `${chosen.storyId}:${chosen.id}` : chosen.id
  );

  const spoken = chosen.answers
    ? answerFor(guest, chosen)
    : `${guest.name} answers, and nothing stands out.`;

  guest.replies = guest.replies || [];
  guest.replies.push({
    prompt: chosen.prompt,
    spoken,
    source: chosen.source || "generic",
  });

  addLog(state, `You: "${chosen.prompt}"`);
  addLog(state, `${guest.name}: "${spoken}"`);

  if (chosen.source === "radio") {
    addLog(
      state,
      "They have heard the radio. Everyone has heard the radio."
    );
  }

  const hidden = (guest.signs ?? []).filter((s) => !s.revealed);
  if (hidden.length) {
    const sign = hidden[Math.floor(Math.random() * hidden.length)];
    sign.revealed = true;
  }
  return true;
}

/** Turn someone away. Never free. */
export function refuseArrival(state, guest) {
  if (!guest) return false;
  state.waitingGuests = state.waitingGuests.filter((g) => g !== guest);

  const story = state.story;
  if (guest.kind === KIND.traveler) {
    state.reputation = Math.max(0, state.reputation - 4);
    addLog(
      state,
      `Turned away ${guest.name}. They had money and somewhere else to be. (−4 reputation)`
    );
  } else if (guest.kind === KIND.survivor) {
    if (story) story.humanity = Math.max(0, story.humanity - 8);
    state.reputation = Math.max(0, state.reputation - 2);
    addLog(
      state,
      `Turned away ${guest.name}. They did not argue. They just walked back toward the road.`
    );
  } else {
    // The player does not get told they were right. Not yet.
    if (story) story.humanity = Math.max(0, story.humanity - 2);
    if (story) story.threatsRefused = (story.threatsRefused || 0) + 1;
    addLog(state, `Turned away ${guest.name}. They did not ask why.`);
    // Delayed confirmation so the call can be learned from later.
    if (story) {
      story.pendingVindication = story.pendingVindication || [];
      story.pendingVindication.push({ name: guest.name, day: state.day });
    }
  }

  storyHook(state, "turnAway", { guest });
  return true;
}

/**
 * Admitting a wrong one does not fail immediately — it fails at night, so the
 * player connects the consequence to the choice after the fact.
 */
export function armAdmittedThreat(state, guest) {
  if (guest?.kind !== KIND.wrong) return;
  if (!state.story) return;
  state.story.pendingThreats = state.story.pendingThreats || [];
  state.story.pendingThreats.push({
    name: guest.name,
    admittedDay: state.day,
    fireOnDay: state.day + 1,
  });
}

/** Called on day rollover. Resolves what admitting or refusing actually meant. */
export function resolveArrivalConsequences(state) {
  const story = state.story;
  if (!story) return;

  const shelterEra = actIndex(state) >= ACT_ORDER.indexOf("collapse");
  const stillPending = [];

  for (const threat of story.pendingThreats || []) {
    if (state.day < threat.fireOnDay) {
      stillPending.push(threat);
      continue;
    }

    const options = [];

    options.push(() => {
      const target = state.rooms.find((r) => r.status === "occupied");
      if (target) {
        target.status = "needs_inspection";
        target.guestName = null;
        target.dirtLevel = "heavy";
        target.hasHiddenDamage = true;
        addLog(
          state,
          `Room ${target.id} is empty this morning. The bed was not slept in and the window is open from the outside.`
        );
      } else {
        addLog(state, "Something moved through the halls last night. Nothing is missing that you can name.");
      }
    });

    if (state.shelter?.unlocked) {
      options.push(() => {
        const taken = Math.min(getShelterStock(state, "food"), 6);
        state.shelter.stock.food -= taken;
        const water = Math.min(getShelterStock(state, "water"), 5);
        state.shelter.stock.water -= water;
        addLog(
          state,
          `Stores were opened overnight. ${taken} food and ${water} water gone. The lock was not forced.`
        );
      });
    }

    if (state.shelter?.defenseActive) {
      options.push(() => {
        state.shelter.integrity = Math.max(0, state.shelter.integrity - 30);
        addLog(
          state,
          `A barricade was dismantled from the inside overnight (${Math.round(state.shelter.integrity)}%).`
        );
      });
    }

    pick(options)();
    addLog(state, `You think about ${threat.name}, and when you checked them in.`);
    if (shelterEra) story.humanity = Math.max(0, story.humanity - 4);
    story.tension += 10;
  }
  story.pendingThreats = stillPending;

  // Late confirmation that a refusal was the right call.
  const stillWaitingProof = [];
  for (const proof of story.pendingVindication || []) {
    if (state.day < proof.day + 2) {
      stillWaitingProof.push(proof);
      continue;
    }
    addLog(
      state,
      `Someone found what was left of ${proof.name} out past the treeline. It had not been a person for a while.`
    );
    story.humanity = Math.min(100, story.humanity + 3);
  }
  story.pendingVindication = stillWaitingProof;
}

File: Unity/Assets/Scripts/Systems/Arrivals.cs
using System;
using System.Collections.Generic;

namespace Vacancy
{
    public static class GuestKind
    {
        public const string Traveler = "traveler";
        public const string Survivor = "survivor";
        public const string Wrong = "wrong";
    }

    public sealed class ArrivalAssessment
    {
        public int BunksFree;
        public int BunksTotal;
        public int Occupants;
        public int SignsRevealed;
        public int QuestionsLeft;
        public ShelterAssessment Shelter;
        public int Humanity;
        public bool PaysRent;
    }

    public sealed class ShelterAssessment
    {
        public int WaterDays;
        public int WaterDaysAfter;
        public int FoodDays;
        public int FoodDaysAfter;
        public int Integrity;
        public bool Powered;
    }

    public static class Arrivals
    {
        const float QuestionHours = 0.75f;

        static readonly string[] Stage1Claims =
        {
            "Driving through. Needs a bed and an early checkout.",
            "Says the next motel is another two hours and they are done driving.",
            "Here for work in the city. Paying by the night.",
            "Asked about the weekly rate, then decided on one night."
        };

        static readonly string[] Stage1Signs =
        {
            "They set a suitcase down and keep a hand on the handle.",
            "They count the cash twice before they speak.",
            "They ask what time the office opens in the morning."
        };

        static readonly Dictionary<string, string[]> Claims = new Dictionary<string, string[]>
        {
            {
                GuestKind.Traveler,
                new[]
                {
                    "Driving through. Needs a bed and an early checkout.",
                    "Says the next motel is another two hours and they are done driving.",
                    "Here for work in the city. Paying by the night.",
                    "Wants the quietest room you have. Does not say why."
                }
            },
            {
                GuestKind.Survivor,
                new[]
                {
                    "Walked here. Asks how many people are already inside.",
                    "Came from the east side. Will not talk about what they saw.",
                    "Has a child with them. Asks only for somewhere with a door.",
                    "Offers to work for a bunk. No money left."
                }
            },
            {
                GuestKind.Wrong,
                new[]
                {
                    "Says they have a reservation. There is no reservation.",
                    "Asks how many people are inside. Asks again, differently.",
                    "Says they came from the city. The roads from the city are closed.",
                    "Wants a room on the ground floor, near the back."
                }
            }
        };

        static readonly string[] DamningSigns =
        {
            "Their reflection in the lobby glass lags a half-second behind them.",
            "They sign the register with the wrong year and do not correct it.",
            "They repeat your last words back to you, quietly, before answering.",
            "They do not blink while you explain the checkout time.",
            "Their bag is empty. You can tell by the way it hangs.",
            "They know Mary's name. Mary has not said it.",
            "They are not breathing when they think you are not looking."
        };

        static readonly string[] InnocuousSigns =
        {
            "They are dressed for winter. It is not winter.",
            "They ask for a room facing away from the road.",
            "They pay entirely in coins.",
            "They flinch at the ice machine.",
            "They will not put their bag down.",
            "They ask twice whether the doors lock from the inside.",
            "They keep checking the window behind you.",
            "They have not slept. It shows."
        };

        public static readonly MediaQuestion[] GenericQuestions =
        {
            new MediaQuestion
            {
                Id = "register-name",
                Source = "generic",
                Prompt = "What name should I put on the register?",
                Answers = new Dictionary<string, string>
                {
                    { GuestKind.Traveler, "The name on my license is fine." },
                    { GuestKind.Survivor, "Just my first name. I do not want it written down twice." },
                    { GuestKind.Wrong, "Any name. Names are for registers." }
                }
            },
            new MediaQuestion
            {
                Id = "how-long",
                Source = "generic",
                Prompt = "How long do you plan to stay?",
                Answers = new Dictionary<string, string>
                {
                    { GuestKind.Traveler, "One night. Maybe two if the roads are still bad." },
                    { GuestKind.Survivor, "Until it is safe. I do not know when that is." },
                    { GuestKind.Wrong, "As long as you will let me. I can stay in any room." }
                }
            },
            new MediaQuestion
            {
                Id = "how-pay",
                Source = "generic",
                Prompt = "How are you paying?",
                Answers = new Dictionary<string, string>
                {
                    { GuestKind.Traveler, "Cash. One night up front, same as the last place." },
                    { GuestKind.Survivor, "I do not have money. I can work if you need hands." },
                    { GuestKind.Wrong, "I can pay. I have what people pay with." }
                }
            },
            new MediaQuestion
            {
                Id = "where-from",
                Source = "generic",
                Prompt = "Where did you come from?",
                Answers = new Dictionary<string, string>
                {
                    { GuestKind.Traveler, "Down the access road. I have been driving since morning." },
                    { GuestKind.Survivor, "On foot. I do not want to say the last town out loud." },
                    { GuestKind.Wrong, "The city. Everyone comes from the city." }
                }
            }
        };

        static T Pick<T>(IList<T> list)
        {
            return list[GameRng.NextInt(0, list.Count - 1)];
        }

        static string PickKind(GameState state)
        {
            if (Stage.IsStageOne(state)) return GuestKind.Traveler;
            int act = Story.ActIndex(state);
            float roll = GameRng.NextFloat();
            int unease = Array.IndexOf(Story.ActOrder, "unease");
            int disruption = Array.IndexOf(Story.ActOrder, "disruption");

            if (act <= unease)
            {
                if (roll < (act == 0 ? 0.05f : 0.14f)) return GuestKind.Wrong;
                return GuestKind.Traveler;
            }

            if (act == disruption)
            {
                if (roll < 0.24f) return GuestKind.Wrong;
                if (roll < 0.62f) return GuestKind.Survivor;
                return GuestKind.Traveler;
            }

            if (roll < 0.3f) return GuestKind.Wrong;
            return GuestKind.Survivor;
        }

        static List<GuestSign> BuildSigns(string kind)
        {
            var signs = new List<GuestSign>();
            var damningPool = new List<string>(DamningSigns);
            var innocuousPool = new List<string>(InnocuousSigns);

            void Take(List<string> pool, bool damning)
            {
                if (pool.Count == 0) return;
                int index = GameRng.NextInt(0, pool.Count - 1);
                string text = pool[index];
                pool.RemoveAt(index);
                signs.Add(new GuestSign { Text = text, Damning = damning, Revealed = false });
            }

            if (kind == GuestKind.Wrong)
            {
                int damningCount = 1 + GameRng.NextInt(0, 2);
                for (int i = 0; i < damningCount; i++) Take(damningPool, true);
                if (GameRng.NextFloat() < 0.5f) Take(innocuousPool, false);
            }
            else
            {
                int innocuousCount = GameRng.NextInt(0, 2);
                for (int i = 0; i < innocuousCount; i++) Take(innocuousPool, false);
                if (GameRng.NextFloat() < 0.16f) Take(damningPool, true);
            }

            for (int i = signs.Count - 1; i > 0; i--)
            {
                int j = GameRng.NextInt(0, i);
                var tmp = signs[i];
                signs[i] = signs[j];
                signs[j] = tmp;
            }

            if (signs.Count > 0) signs[0].Revealed = true;
            return signs;
        }

        static List<GuestSign> BuildStage1Signs()
        {
            var signs = new List<GuestSign>();
            if (GameRng.NextFloat() >= 0.45f) return signs;
            signs.Add(new GuestSign
            {
                Text = Pick(Stage1Signs),
                Damning = false,
                Revealed = true
            });
            return signs;
        }

        public static WaitingGuest CreateArrival(GameState state, string name)
        {
            if (Stage.IsStageOne(state))
            {
                var ordinary = new WaitingGuest
                {
                    Name = name,
                    Kind = GuestKind.Traveler,
                    StoryId = null,
                    Claim = Pick(Stage1Claims),
                    WaitRemainingHours = GameConfig.WaitPatienceHours,
                    Marked = false
                };
                ordinary.Signs.AddRange(BuildStage1Signs());
                return ordinary;
            }

            string kind = PickKind(state);
            var story = Media.PickTiedStory(state, kind);
            var guest = new WaitingGuest
            {
                Name = name,
                Kind = kind,
                StoryId = story?.Id,
                Claim = Pick(Claims[kind]),
                WaitRemainingHours = GameConfig.WaitPatienceHours,
                Marked = kind != GuestKind.Traveler
            };
            guest.Signs.AddRange(BuildSigns(kind));
            return guest;
        }

        public static List<MediaQuestion> DeskQuestions(GameState state, WaitingGuest guest)
        {
            if (Stage.IsStageOne(state))
            {
                var askedIds = new HashSet<string>(guest?.AskedQuestionIds ?? new List<string>());
                var basics = new List<MediaQuestion>();
                foreach (var q in GenericQuestions)
                {
                    if (!askedIds.Contains(q.Id)) basics.Add(q);
                }

                return basics;
            }

            var mediaQs = Media.AvailableQuestions(state, guest);
            var asked = new HashSet<string>(guest?.AskedQuestionIds ?? new List<string>());
            var list = new List<MediaQuestion>();
            foreach (var q in mediaQs)
            {
                string key = q.StoryId != null ? $"{q.StoryId}:{q.Id}" : q.Id;
                if (!asked.Contains(key)) list.Add(q);
            }

            foreach (var q in GenericQuestions)
            {
                if (!asked.Contains(q.Id)) list.Add(q);
            }

            return list;
        }

        public static List<GuestSign> RevealedSigns(WaitingGuest guest)
        {
            var list = new List<GuestSign>();
            if (guest?.Signs == null) return list;
            foreach (var sign in guest.Signs)
            {
                if (sign.Revealed) list.Add(sign);
            }

            return list;
        }

        public static int HiddenSignCount(WaitingGuest guest)
        {
            int count = 0;
            if (guest?.Signs == null) return 0;
            foreach (var sign in guest.Signs)
            {
                if (!sign.Revealed) count++;
            }

            return count;
        }

        public static bool HasVisibleTell(WaitingGuest guest)
        {
            foreach (var sign in RevealedSigns(guest))
            {
                if (sign.Damning) return true;
            }

            return false;
        }

        public static ArrivalAssessment AssessArrival(GameState state, WaitingGuest guest)
        {
            int bunksFree = 0;
            int bunksTotal = 0;
            foreach (var room in state.Rooms)
            {
                if (!room.Unlocked) continue;
                bunksTotal++;
                if (room.Status == "clean") bunksFree++;
            }

            var assessment = new ArrivalAssessment
            {
                BunksFree = bunksFree,
                BunksTotal = bunksTotal,
                Occupants = Shelter.CountOccupants(state),
                SignsRevealed = RevealedSigns(guest).Count,
                QuestionsLeft = Math.Max(0, guest.MaxQuestions - guest.QuestionsAsked),
                Humanity = state.Story?.Humanity ?? 100,
                PaysRent = guest.Kind == GuestKind.Traveler
            };

            if (state.Shelter != null && state.Shelter.Unlocked && !Stage.IsStageOne(state))
            {
                int now = assessment.Occupants;
                int after = now + 1;
                int Days(int stock, float perPerson)
                {
                    return perPerson > 0 ? (int)Math.Floor(stock / (perPerson * Math.Max(1, now))) : 0;
                }

                int DaysAfter(int stock, float perPerson)
                {
                    return perPerson > 0 ? (int)Math.Floor(stock / (perPerson * Math.Max(1, after))) : 0;
                }

                assessment.Shelter = new ShelterAssessment
                {
                    WaterDays = Days(Shelter.GetStock(state, "water"), GameConfig.WaterPerPersonPerDay),
                    WaterDaysAfter = DaysAfter(Shelter.GetStock(state, "water"), GameConfig.WaterPerPersonPerDay),
                    FoodDays = Days(Shelter.GetStock(state, "food"), GameConfig.FoodPerPersonPerDay),
                    FoodDaysAfter = DaysAfter(Shelter.GetStock(state, "food"), GameConfig.FoodPerPersonPerDay),
                    Integrity = (int)Math.Round(state.Shelter.Integrity),
                    Powered = state.Shelter.Powered
                };
            }

            return assessment;
        }

        public static bool AskQuestion(GameState state, WaitingGuest guest, MediaQuestion question = null)
        {
            if (guest == null) return false;
            if (guest.QuestionsAsked >= guest.MaxQuestions)
            {
                state.AddLog($"{guest.Name} stops answering questions.");
                return false;
            }

            var options = DeskQuestions(state, guest);
            var chosen = question ?? (options.Count > 0 ? options[0] : GenericQuestions[0]);
            guest.QuestionsAsked += 1;
            guest.WaitRemainingHours -= QuestionHours;
            state.Hour += QuestionHours;
            guest.AskedQuestionIds.Add(chosen.StoryId != null ? $"{chosen.StoryId}:{chosen.Id}" : chosen.Id);

            string spoken = chosen.Answers != null
                ? Media.AnswerFor(guest, chosen)
                : $"{guest.Name} answers, and nothing stands out.";

            guest.Replies.Add(new GuestReply
            {
                Prompt = chosen.Prompt,
                Spoken = spoken,
                Source = chosen.Source ?? "generic"
            });

            state.AddLog($"You: \"{chosen.Prompt}\"");
            state.AddLog($"{guest.Name}: \"{spoken}\"");

            if (chosen.Source == "radio")
            {
                state.AddLog("They have heard the radio. Everyone has heard the radio.");
            }

            var hidden = new List<GuestSign>();
            foreach (var sign in guest.Signs)
            {
                if (!sign.Revealed) hidden.Add(sign);
            }

            if (hidden.Count > 0)
            {
                hidden[GameRng.NextInt(0, hidden.Count - 1)].Revealed = true;
            }

            return true;
        }

        public static bool RefuseArrival(GameState state, WaitingGuest guest)
        {
            if (guest == null) return false;
            state.WaitingGuests.Remove(guest);

            var story = state.Story;
            if (guest.Kind == GuestKind.Traveler)
            {
                state.Reputation = Math.Max(0, state.Reputation - 4);
                state.AddLog(
                    $"Turned away {guest.Name}. They had money and somewhere else to be. (−4 reputation)");
            }
            else if (guest.Kind == GuestKind.Survivor)
            {
                if (story != null) story.Humanity = Math.Max(0, story.Humanity - 8);
                state.Reputation = Math.Max(0, state.Reputation - 2);
                state.AddLog(
                    $"Turned away {guest.Name}. They did not argue. They just walked back toward the road.");
            }
            else
            {
                if (story != null)
                {
                    story.Humanity = Math.Max(0, story.Humanity - 2);
                    story.ThreatsRefused += 1;
                    story.PendingVindication.Add(new PendingVindication { Name = guest.Name, Day = state.Day });
                }

                state.AddLog($"Turned away {guest.Name}. They did not ask why.");
            }

            Story.Hook(state, "turnAway", guest);
            return true;
        }

        public static void ArmAdmittedThreat(GameState state, WaitingGuest guest)
        {
            if (guest?.Kind != GuestKind.Wrong) return;
            if (state.Story == null) return;
            state.Story.PendingThreats.Add(new PendingThreat
            {
                Name = guest.Name,
                AdmittedDay = state.Day,
                FireOnDay = state.Day + 1
            });
        }

        public static void ResolveConsequences(GameState state)
        {
            var story = state.Story;
            if (story == null) return;

            bool shelterEra = Story.IsShelterEra(state);
            var stillPending = new List<PendingThreat>();

            foreach (var threat in story.PendingThreats)
            {
                if (state.Day < threat.FireOnDay)
                {
                    stillPending.Add(threat);
                    continue;
                }

                var options = new List<Action>();
                options.Add(() =>
                {
                    Room target = null;
                    foreach (var room in state.Rooms)
                    {
                        if (room.Status == "occupied")
                        {
                            target = room;
                            break;
                        }
                    }

                    if (target != null)
                    {
                        target.Status = "needs_inspection";
                        target.GuestName = null;
                        target.DirtLevel = "heavy";
                        target.HasHiddenDamage = true;
                        state.AddLog(
                            $"Room {target.Id} is empty this morning. The bed was not slept in and the window is open from the outside.");
                    }
                    else
                    {
                        state.AddLog("Something moved through the halls last night. Nothing is missing that you can name.");
                    }
                });

                if (state.Shelter != null && state.Shelter.Unlocked)
                {
                    options.Add(() =>
                    {
                        int taken = Math.Min(Shelter.GetStock(state, "food"), 6);
                        state.Shelter.Stock["food"] -= taken;
                        int water = Math.Min(Shelter.GetStock(state, "water"), 5);
                        state.Shelter.Stock["water"] -= water;
                        state.AddLog(
                            $"Stores were opened overnight. {taken} food and {water} water gone. The lock was not forced.");
                    });
                }

                if (state.Shelter != null && state.Shelter.DefenseActive)
                {
                    options.Add(() =>
                    {
                        state.Shelter.Integrity = Math.Max(0, state.Shelter.Integrity - 30);
                        state.AddLog(
                            $"A barricade was dismantled from the inside overnight ({Math.Round(state.Shelter.Integrity)}%).");
                    });
                }

                Pick(options)();
                state.AddLog($"You think about {threat.Name}, and when you checked them in.");
                if (shelterEra) story.Humanity = Math.Max(0, story.Humanity - 4);
                story.Tension += 10;
            }

            story.PendingThreats.Clear();
            story.PendingThreats.AddRange(stillPending);

            var stillWaitingProof = new List<PendingVindication>();
            foreach (var proof in story.PendingVindication)
            {
                if (state.Day < proof.Day + 2)
                {
                    stillWaitingProof.Add(proof);
                    continue;
                }

                state.AddLog(
                    $"Someone found what was left of {proof.Name} out past the treeline. It had not been a person for a while.");
                story.Humanity = Math.Min(100, story.Humanity + 3);
            }

            story.PendingVindication.Clear();
            story.PendingVindication.AddRange(stillWaitingProof);
        }
    }
}
